internal/runtime/core.go

```diff
@@ -44,18 +47,63 @@ func NewServiceAccount(namespace, name string) *corev1.ServiceAccount {
 }
 
 // NewSecret returns a corev1.Secret with namespace and name.
-func NewSecret(namespace, name string, data map[string][]byte) *corev1.Secret {
+func NewSecret(namespace, name string, data map[string][]byte, visitors ...func(meta metav1.Object)) *corev1.Secret {
 	if data == nil {
 		data = map[string][]byte{}
 	}
 	s := &corev1.Secret{Data: data}
-	Initialize(s, namespace, name)
+	Initialize(s, namespace, name, visitors...)
 	return s
 }
 
-// NewDaemonSet returns a daemon set.
-func NewDaemonSet(namespace, name string) *appsv1.DaemonSet {
-	ds := &appsv1.DaemonSet{}
-	Initialize(ds, namespace, name)
+func NewDaemonSet(daemonSetName, namespace, component string, podSpec corev1.PodSpec, visitors ...func(meta metav1.Object)) *appsv1.DaemonSet {
+	labels := map[string]string{
+		"provider":      "openshift",
+		"component":     component,
+		"logging-infra": component,
+	}
+	ds := &appsv1.DaemonSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DaemonSet",
+			APIVersion: appsv1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      daemonSetName,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   daemonSetName,
+					Labels: labels,
+					Annotations: map[string]string{
+						"scheduler.alpha.kubernetes.io/critical-pod": "",
+						"target.workload.openshift.io/management":    `{"effect": "PreferredDuringScheduling"}`,
+					},
+				},
+				Spec: podSpec,
+			},
+			UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+				Type: appsv1.RollingUpdateDaemonSetStrategyType,
+				RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+					MaxUnavailable: &intstr.IntOrString{
+						Type:   intstr.String,
+						StrVal: "100%",
+					},
+				},
+			},
+		},
+	}
+	Initialize(ds, namespace, daemonSetName, visitors...)
+	//TODO: should we keep this labels?
+	utils.AddLabels(ds, labels)
+
+	dl := ds.GetLabels()
+	ds.Spec.Template.SetLabels(dl)
```

Comment:
I would expect this to be encapsulated into a visitor as ell

Reply:
@jcantrill 
Transferring this operation to the visitor will require of identifying the type of K8s object, as the parameter involved is a `runtime.Object`, e.g:
```
gvk := runtime.GroupVersionKind(o)
if gvk.Group == "apps" && gvk.Kind == "DaemonSet" {
    ds := o.(*apps.DaemonSet)
    utils.SetCommonLabelsDaemonSet(ds)
} else {
    utils.SetCommonLabels(o)
}
```
This is the reason why I prefer not to move this logic to the visitor.

Reply:
Looks like the context for these comments has changed, I don't see a `SetCommonLabelsDaemonSet` anymore.